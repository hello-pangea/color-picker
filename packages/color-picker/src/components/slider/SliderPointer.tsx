import React from "react";

export default function SliderPointer() {
  const styles: Record<string, React.CSSProperties> = {
    picker: {
      width: "14px",
      height: "14px",
      borderRadius: "6px",
      transform: "translate(-7px, -1px)",
      backgroundColor: "rgb(248, 248, 248)",
      boxShadow: "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
    },
  };

  return <div style={styles.picker} />;
}
