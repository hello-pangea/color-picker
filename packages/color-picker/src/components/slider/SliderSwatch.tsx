import React from "react";
import { HslColor } from "../../types/colors";

type Props = {
  hsl: HslColor;
  onClick: any;
  offset: number;
  active?: boolean;
  first?: boolean;
  last?: boolean;
};

export default function SliderSwatch({
  hsl,
  offset,
  onClick = () => {},
  active,
  first,
  last,
}: Props) {
  const styles: Record<string, React.CSSProperties> = {
    swatch: {
      height: "12px",
      background: `hsl(${hsl.h}, 50%, ${offset * 100}%)`,
      cursor: "pointer",
      borderRadius: active
        ? "3.6px/2px"
        : first
        ? "2px 0 0 2px"
        : last
        ? "0 2px 2px 0"
        : undefined,
      transform: active ? "scaleY(1.8)" : undefined,
    },
  };

  const handleClick = (e: React.MouseEvent) =>
    onClick(
      {
        h: hsl.h,
        s: 0.5,
        l: offset,
        source: "hsl",
      },
      e
    );

  return <div style={styles.swatch} onClick={handleClick} />;
}
