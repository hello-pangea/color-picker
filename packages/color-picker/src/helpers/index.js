export * from "./checkboard";
export * from "./color";
