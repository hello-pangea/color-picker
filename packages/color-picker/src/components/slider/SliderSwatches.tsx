import React from "react";
import { HslColor } from "../../types/colors";
import SliderSwatch from "./SliderSwatch";

type Props = {
  onClick: any;
  hsl: HslColor;
};

export default function SliderSwatches({ onClick, hsl }: Props) {
  const styles: Record<string, React.CSSProperties> = {
    swatches: {
      marginTop: "20px",
    },
    swatch: {
      boxSizing: "border-box",
      width: "20%",
      paddingRight: "1px",
      float: "left",
    },
    clear: {
      clear: "both",
    },
  };

  // Acceptible difference in floating point equality
  const epsilon = 0.1;

  return (
    <div style={styles.swatches}>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.8}
          active={
            Math.abs(hsl.l - 0.8) < epsilon && Math.abs(hsl.s - 0.5) < epsilon
          }
          onClick={onClick}
          first
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.65}
          active={
            Math.abs(hsl.l - 0.65) < epsilon && Math.abs(hsl.s - 0.5) < epsilon
          }
          onClick={onClick}
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.5}
          active={
            Math.abs(hsl.l - 0.5) < epsilon && Math.abs(hsl.s - 0.5) < epsilon
          }
          onClick={onClick}
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.35}
          active={
            Math.abs(hsl.l - 0.35) < epsilon && Math.abs(hsl.s - 0.5) < epsilon
          }
          onClick={onClick}
        />
      </div>
      <div style={styles.swatch}>
        <SliderSwatch
          hsl={hsl}
          offset={0.2}
          active={
            Math.abs(hsl.l - 0.2) < epsilon && Math.abs(hsl.s - 0.5) < epsilon
          }
          onClick={onClick}
          last
        />
      </div>
      <div style={styles.clear} />
    </div>
  );
}
