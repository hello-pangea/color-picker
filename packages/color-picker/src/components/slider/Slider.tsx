import merge from "lodash/merge";
import React from "react";
import { useColor, withColorProvider } from "../../context/useColor";
import { Hue } from "../common";
import SliderPointer from "./SliderPointer";
import SliderSwatches from "./SliderSwatches";

export type SliderPickerProps = {
  pointer?: typeof SliderPointer;
  styles?: Record<string, React.CSSProperties>;
  className?: string;
};

const Slider = ({
  pointer = SliderPointer,
  styles: passedStyles = {},
  className = "",
}: SliderPickerProps) => {
  const { colors, changeColor } = useColor();
  const { hsl } = colors;

  const styles = merge<
    Record<string, React.CSSProperties>,
    Record<string, React.CSSProperties>
  >(
    {
      hue: {
        height: "12px",
        position: "relative",
      },
      Hue: {
        borderRadius: "2px",
      },
    },
    passedStyles
  );

  return (
    <div style={styles.wrap || {}} className={`slider-picker ${className}`}>
      <div style={styles.hue}>
        <Hue
          style={styles.Hue}
          hsl={hsl}
          pointer={pointer}
          onChange={changeColor}
        />
      </div>
      <div style={styles.swatches}>
        <SliderSwatches hsl={hsl} onClick={changeColor} />
      </div>
    </div>
  );
};

export default withColorProvider(Slider);
