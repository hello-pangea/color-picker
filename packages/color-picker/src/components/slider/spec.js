/* global test, expect */

import React from "react";
import renderer from "react-test-renderer";
import { red } from "../../helpers/color";

import Slider from "./Slider";
import SliderPointer from "./SliderPointer";
import SliderSwatch from "./SliderSwatch";
import SliderSwatches from "./SliderSwatches";

test("Slider renders correctly", () => {
  const tree = renderer.create(<Slider {...red} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Slider renders custom styles correctly", () => {
  const tree = renderer
    .create(<Slider styles={{ default: { wrap: { boxShadow: "none" } } }} />)
    .toJSON();
  expect(tree.props.style.boxShadow).toBe("none");
});

test("SliderPointer renders correctly", () => {
  const tree = renderer.create(<SliderPointer />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("SliderSwatch renders correctly", () => {
  const tree = renderer.create(<SliderSwatch {...red} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("SliderSwatches renders correctly", () => {
  const tree = renderer.create(<SliderSwatches {...red} />).toJSON();
  expect(tree).toMatchSnapshot();
});
