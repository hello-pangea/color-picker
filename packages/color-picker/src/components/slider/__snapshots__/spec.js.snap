// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders correctly 1`] = `
<div
  className="slider-picker "
  style={{}}
>
  <div
    style={
      {
        "height": "12px",
        "position": "relative",
      }
    }
  >
    <div
      style={
        {
          "borderRadius": undefined,
          "boxShadow": undefined,
          "inset": "0px",
          "position": "absolute",
        }
      }
    >
      <div
        className="hue-horizontal"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "borderRadius": undefined,
            "height": "100%",
            "padding": "0 2px",
            "position": "relative",
          }
        }
      >
        <style>
          
            .hue-horizontal {
              background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }

            .hue-vertical {
              background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }
          
        </style>
        <div
          style={
            {
              "left": "69.44444444444443%",
              "position": "absolute",
              "top": undefined,
            }
          }
        >
          <div
            style={
              {
                "backgroundColor": "rgb(248, 248, 248)",
                "borderRadius": "6px",
                "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "height": "14px",
                "transform": "translate(-7px, -1px)",
                "width": "14px",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      style={
        {
          "marginTop": "20px",
        }
      }
    >
      <div
        style={
          {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            {
              "background": "hsl(249.99999999999994, 50%, 80%)",
              "borderRadius": "2px 0 0 2px",
              "cursor": "pointer",
              "height": "12px",
              "transform": undefined,
            }
          }
        />
      </div>
      <div
        style={
          {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            {
              "background": "hsl(249.99999999999994, 50%, 65%)",
              "borderRadius": undefined,
              "cursor": "pointer",
              "height": "12px",
              "transform": undefined,
            }
          }
        />
      </div>
      <div
        style={
          {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            {
              "background": "hsl(249.99999999999994, 50%, 50%)",
              "borderRadius": undefined,
              "cursor": "pointer",
              "height": "12px",
              "transform": undefined,
            }
          }
        />
      </div>
      <div
        style={
          {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            {
              "background": "hsl(249.99999999999994, 50%, 35%)",
              "borderRadius": undefined,
              "cursor": "pointer",
              "height": "12px",
              "transform": undefined,
            }
          }
        />
      </div>
      <div
        style={
          {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            {
              "background": "hsl(249.99999999999994, 50%, 20%)",
              "borderRadius": "3.6px/2px",
              "cursor": "pointer",
              "height": "12px",
              "transform": "scaleY(1.8)",
            }
          }
        />
      </div>
      <div
        style={
          {
            "clear": "both",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`SliderPointer renders correctly 1`] = `
<div
  style={
    {
      "backgroundColor": "rgb(248, 248, 248)",
      "borderRadius": "6px",
      "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "height": "14px",
      "transform": "translate(-7px, -1px)",
      "width": "14px",
    }
  }
/>
`;

exports[`SliderSwatch renders correctly 1`] = `
<div
  onClick={[Function]}
  style={
    {
      "background": "hsl(0, 50%, NaN%)",
      "borderRadius": undefined,
      "cursor": "pointer",
      "height": "12px",
      "transform": undefined,
    }
  }
/>
`;

exports[`SliderSwatches renders correctly 1`] = `
<div
  style={
    {
      "marginTop": "20px",
    }
  }
>
  <div
    style={
      {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        {
          "background": "hsl(0, 50%, 80%)",
          "borderRadius": "2px 0 0 2px",
          "cursor": "pointer",
          "height": "12px",
          "transform": undefined,
        }
      }
    />
  </div>
  <div
    style={
      {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        {
          "background": "hsl(0, 50%, 65%)",
          "borderRadius": undefined,
          "cursor": "pointer",
          "height": "12px",
          "transform": undefined,
        }
      }
    />
  </div>
  <div
    style={
      {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        {
          "background": "hsl(0, 50%, 50%)",
          "borderRadius": undefined,
          "cursor": "pointer",
          "height": "12px",
          "transform": undefined,
        }
      }
    />
  </div>
  <div
    style={
      {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        {
          "background": "hsl(0, 50%, 35%)",
          "borderRadius": undefined,
          "cursor": "pointer",
          "height": "12px",
          "transform": undefined,
        }
      }
    />
  </div>
  <div
    style={
      {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        {
          "background": "hsl(0, 50%, 20%)",
          "borderRadius": "0 2px 2px 0",
          "cursor": "pointer",
          "height": "12px",
          "transform": undefined,
        }
      }
    />
  </div>
  <div
    style={
      {
        "clear": "both",
      }
    }
  />
</div>
`;
